lib/rubocop/ast/node_pattern/compiler.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module AST
+    class NodePattern
+      # The top-level compiler holding the global state
+      # Defers work to its subcompilers
+      #
+      # Doc on how this fits in the compiling process:
+      #   /doc/modules/ROOT/pages/node_pattern.md
+      class Compiler
+        extend Forwardable
+        attr_reader :captures, :named_parameters, :positional_parameters, :binding
+
+        def initialize
+          @temp_depth = 0 # avoid name clashes between temp variables
+          @captures = 0 # number of captures seen
+          @positional_parameters = 0 # highest % (param) number seen
+          @named_parameters = Set[] # keyword parameters
+          @binding = Binding.new # bound variables
+          @atom_subcompiler = self.class::AtomSubcompiler.new(self)
```

Comment:
Is `self.class::` necessary? Removed all occurences in this file, and all specs passed.

Reply:
This one isn't, because the `AtomSubcompiler` isn't refined by any subclasses... It might in the future though, so I'd rather keep it this way.